File: random_forest/python/rf.py
import random

NUM_TREES = 10
NUM_FEATURES = 5

def load_data(filename):
    data = []
    with open(filename, 'r') as f:
        for line in f:
            features = line.split(',')
            label = int(features[0])
            feature_values = [float(f) for f in features[1:]]
            data.append({'label': label, 'features': feature_values})
    return data

def create_random_forest():
    tree_labels = []
    for _ in range(NUM_TREES):
        tree_labels.append(random.randint(0, 1))
    return tree_labels

def classify_data(data, tree_labels):
    predictions = []
    for dp in data:
        correct_label = dp['label']
        prediction = 0
        for tree_label in tree_labels:
            if tree_label == correct_label:
                prediction += 1
        predictions.append(prediction > NUM_TREES / 2)
    return predictions

def calculate_accuracy(predictions, data):
    correct = sum(predictions == dp['label'] for dp in data)
    accuracy = correct / len(data)
    return accuracy * 100

def main():
    data = load_data('data.csv')
    tree_labels = create_random_forest()
    predictions = classify_data(data, tree_labels)
    accuracy = calculate_accuracy(predictions, data)
    print('Accuracy: {}%'.format(accuracy))

if __name__ == '__main__':
    main()

File: random_forest/julia/rf.jl
using Random

const NUM_TREES = 10
const NUM_FEATURES = 5

function load_data(filename)
  data = []
  open(filename) do f
    for line in readlines(f)
      features = split(line, ",")
      label = parse(Int, features[1])
      feature_values = [parse(Float64, f) for f in features[2:end]]
      data.push({label, feature_values})
    end
  end
  data
end

function create_random_forest()
  tree_labels = rand(Bool, NUM_TREES)
  tree_labels
end

function classify_data(data, tree_labels)
  predictions = []
  for dp in data
    correct_label = dp.label
    prediction = 0
    for tree_label in tree_labels
      if tree_label == correct_label
        prediction += 1
      end
    end
    predictions.push(prediction > NUM_TREES / 2 ? 1 : 0)
  end
  predictions
end

function calculate_accuracy(predictions, data)
  correct = sum(predictions .== data.label)
  accuracy = correct / length(predictions)
  accuracy * 100
end

function main()
  data = load_data("data.csv")
  tree_labels = create_random_forest()
  predictions = classify_data(data, tree_labels)
  accuracy = calculate_accuracy(predictions, data)
  println("Accuracy: ", accuracy, "%")
end

main()
